docs/proposals/devfile/outer-loop-build-and-deploy.md

```diff
@@ -0,0 +1,310 @@
+
+We will cover two main stages:
+1. Build the image
+2. Deployment of the image. Deploy the image to a cluster
+
+## Build the image
+
+This includes the creation of the base image of the runtime container, building the application, and packaging the application as part of the container image. Different mechanisms can be used for building the image, for example:
+1. Dockerfile
+2. Dockerfile stages
+3. Buildpacks
+4. Source-to-Image (S2I)
+5. Tekton
+
+The initial design will focus on the first two items for discussion. Other ways of building the image will be covered later.
+
+The definition of the image build introduces a new component type `image` for the definition of the image build. The `image` component definition can use different build strategies, e.g. `dockerfile`, `s2i`, `buildpacks`, etc.
+
+__Alternative approach:__ we can consider reusing the `exec` type except that these build types are special and we may not want to incorporate the full image build command into an `exec` command. Therefore, I chose to use a different type of command here.
+
+### Example using a file Dockerfile as part of the app:
+    variables:
+      myimage: myimagename
+    components:
+      - name: mydockerfileimage
+        image:
+          imageName: {{myimage}}
+          dockerfile:
+            buildContext: ${PROJECTS_ROOT}/build
+            location: Dockerfile
+            args: [ "arg1", "arg2", "arg3" ]
+            rootRequired: false
+
+`imageName`: name of the generated image (or we may use an id for referencing the image). This imageName can also be fully qualified to allow pushing of the image to a specific image registry.
+
+`buildContext`: Path of source directory to establish build context.  Default to ${PROJECT_ROOT} (optional)
+
+`location`: Dockerfile location which can be an URL or a path relative to buildContext
+
+`args`: Argument list for the docker build (optional)
+
+`rootRequired`: Specifies whether a privileged builder pod is required.  Default is false. (optional)
+
+__Note:__
+1. A common pattern will be using a global variable to define the image name (`myimage` in the example above) so that it can be easily referred to in the deploy step later.
+
+### Example using a Dockerfile with registry and secret for the image push:
+    variables:
+      myimage: myimagename
+    components:
+      - name: mydockerfileimage
+        image:
+          imageName: {{myimage}}
+          dockerfile:
+            buildContext: ${PROJECTS_ROOT}/build
+            location: https://github.com/redhat-developer/devfile-sample/blob/master/src/Dockerfile
+            args: [ "arg1", "arg2", "arg3" ]
+            rootRequired: false
+            envFrom:
+            - secretRef:
+                name: my-secret
```

Comment:
Enclose in ` ```yaml (....) ``` `

Applies to other code blocks in this document as well.

Reply:
fixed all